### ISSUE TITLE
Cannot enable GPU with New Backend and Python3

### ISSUE THREAD

**Issue by pnavarre** [Theano/Theano#5484]
Hi,
I cannot enable GPU with new backend, latest git versions of Theano+libgpuarray, when using Python 3. Python 2 works.
Thanks!
Pablo

####
Name: Theano
Version: 0.9.0b1
Summary: Optimizing compiler for evaluating mathematical expressions on CPUs and GPUs.
Home-page: http://deeplearning.net/software/theano/
Author: LISA laboratory, University of Montreal
Author-email: theano-dev@googlegroups.com
License: BSD
Location: /opt/libs/python3_
env/lib/python3.5/site-packagesRequires: numpy, scipy, six

####
GIT version:
Theano 8b9f73365e4932f1c005a0a37b907d28985fbc5f
libgpuarray https://github.com/Theano/libgpuarray/pull/338

####
In Python 3.5.3 or 3.6.0:

```
import os
os.environ['THEANO_FLAGS'] = """
    device=cuda0,
    floatX=float32,
"""
import theano
```

Output:
```
Using cuDNN version 5105 on context None

ERROR (theano.gpuarray): Could not initialize pygpu, support disabled
Traceback (most recent call last):
  File "/opt/libs/python3_env/lib/python3.5/site-packages/theano/gpuarray/__init__.py", line 164, in <module>
    use(config.device)
  File "/opt/libs/python3_env/lib/python3.5/site-packages/theano/gpuarray/__init__.py", line 151, in use
    init_dev(device)
  File "/opt/libs/python3_env/lib/python3.5/site-packages/theano/gpuarray/__init__.py", line 100, in init_dev

    pygpu.blas.gemm(0, tmp, tmp, 0, tmp, overwrite_c=True)
  File "pygpu/blas.pyx", line 129, in pygpu.blas.gemm (pygpu/blas.c:3354)
  File "pygpu/blas.pyx", line 44, in pygpu.blas.pygpu_blas_rgemm (pygpu/blas.c:2011)
pygpu.gpuarray.GpuArrayException: (b'Unsupported operation', 5)
```
####
In Python 2.7.13:

```
import os
os.environ['THEANO_FLAGS'] = """
    device=cuda0,
    floatX=float32,
"""
import theano
```

Output:
```
Using cuDNN version 5105 on context None
Mapped name None to device cuda0: GeForce GTX 960M (0000:01:00.0)
```

**Comment by pnavarre** [comment]
I found that this issue started in https://github.com/Theano/Theano/commit/d133bf69982bf04c93aed7ac3728ccf5ab3ef710.
Version https://github.com/Theano/Theano/commit/e0897668387bb5fd142b995716fd0c5066434ee1 does not have this issue.

In case it helps, I use a virtual environment in Ubuntu 16.04 LTS. Numpy version is 1.12.0.

**Comment by lamblin** [comment]
@pnavarre That commit has been mostly reverted in a subsequent one. Could you try with cb35ac2485d6bc28cba8a7005517cb59db29a15c instead?

**Comment by pnavarre** [comment]
Thank you for checking this issue. Version cb35ac2485d6bc28cba8a7005517cb59db29a15c has the same problem.

I tried Python 3.6.0 and got:
```
Using cuDNN version 5105 on context None
ERROR (theano.gpuarray): Could not initialize pygpu, support disabled
Traceback (most recent call last):
  File "/opt/libs/python3_env/lib/python3.6/site-packages/theano/gpuarray/__init__.py", line 164, in <module>
    use(config.device)
  File "/opt/libs/python3_env/lib/python3.6/site-packages/theano/gpuarray/__init__.py", line 151, in use
    init_dev(device)
  File "/opt/libs/python3_env/lib/python3.6/site-packages/theano/gpuarray/__init__.py", line 100, in init_dev
    pygpu.blas.gemm(0, tmp, tmp, 0, tmp, overwrite_c=True)
  File "pygpu/blas.pyx", line 129, in pygpu.blas.gemm (pygpu/blas.c:3350)
  File "pygpu/blas.pyx", line 44, in pygpu.blas.pygpu_blas_rgemm (pygpu/blas.c:2007)
pygpu.gpuarray.GpuArrayException: (b'Unsupported operation', 5)
```

Python 2.7.13 doesn't have problems.

Same error persists in last git version f3844589c0f24b43d3d7cc22de713196fef1bc4d

**Comment by lamblin** [comment]
You're not the only one to report issues with python 3.6...
Could you try with python 3.5, to help isolate the issue?

**Comment by pnavarre** [comment]
Hi, I got the same error using Python 3.5.3 in latest git version bcf2d7dca71937453f3f6d1b3edddd6d2ad18ef7 as well as cb35ac2485d6bc28cba8a7005517cb59db29a15c.
The issue is very consistent for me. I get it in a laptop with GTX 960M, on a desktop computer with Titan X (Maxwell), and a server with Tesla M40 24GB. All of them use Ubuntu 16.04, and a local environment with CUDA 8.0, cudnn 5.1, libgpuarray 0.6.1.

**Comment by pnavarre** [comment]
Just in case, I also tried different Nvidia drivers: 375.39 and 370.28. Same problem.

**Comment by lamblin** [comment]
Would it be possible for you to share the code reproducing the issue?

**Comment by pnavarre** [comment]
Yes, I just run a script as shown in the original post (with empty .theanorc):

`import os
os.environ['THEANO_FLAGS'] = """
    device=cuda0,
    floatX=float32,
"""
import theano`

This sets up the flags for device, floatX, and then import theano. The error shows soon after import theano.

**Comment by lamblin** [comment]
OK, I have a machine with Ubuntu 16.04 and a GTX 580, I'll try to replicate.

**Comment by lamblin** [comment]
I just tried it:
- Ubuntu 16.04 LTS, updated today
- cuda 8.0 (no cuDNN, did not bother with a 580) with driver 375.26
- Python 3.5.2 (python3 from system), in a virtualenv with numpy 1.12, cython 0.25 (installed by pip in the virtualenv)
- libgpuarray and pygpu just installed, latest master from Theano

And it seemed to work:
```
GPUARRAY_FORCE_CUDA_DRIVER_LOAD=1 python test_5484.py                                          20:37
WARNING: loading blacklisted driver because the load was forced.
Can not use cuDNN on context None: cannot compile with cuDNN. We got this error:
b'/tmp/try_flags_m8c_yhtx.c:4:19: fatal error: cudnn.h: No such file or directory\ncompilation terminated.\n'
Mapped name None to device cuda0: GeForce GTX 580 (0000:02:00.0)
```

**Comment by pnavarre** [comment]
Thanks a lot for testing!
Based on your test I looked at the details of my environment and found something.

As mentioned above, in Python 2 I don't have any problem. Here, I use [virtualenv](http://www.virtualenv.org/) 15.1.0 to generate the environment.

Python 3 has a built-in support for virtual environments based on [PEP 405](https://www.python.org/dev/peps/pep-0405/). See more in this [link](https://docs.python.org/3/library/venv.html). This is what I used to create my environment (both in Python 3.5 and 3.6 tests above). Specifically:
`python3 -m venv /opt/libs/python3_env`

I tried using [virtualenv](http://www.virtualenv.org/) 15.1.0 in Python 3 and the problem of this issue dissapears. So, **the issue is "solved" (avoided) by using [virtualenv](http://www.virtualenv.org/)**.

Please keep in mind that this might still be a problem since in Python 3 and above, [virtualenv](http://www.virtualenv.org/) is basically deprecated according to my understanding. There seems to be technical disadvantages with [virtualenv](http://www.virtualenv.org/). You can find more in Motivation in [PEP 405](https://www.python.org/dev/peps/pep-0405/). I have been using this approach to create virtual environment for months with no problem except this.

**Comment by pnavarre** [comment]
Sorry, I deleted my previous post about conflict related to virtualenv. I need to do more test. Maybe there is something else affecting my virtual environment.

**Comment by pnavarre** [comment]
So far I know that I have a conflict with another library. If I only compile Theano + dependencies I do not get the error. My candidate so far is "pycuda" or one of its dependencies, perhaps "boost", both compiled from source. I will let you know when I confirm the error.

**Comment by pnavarre** [comment]
@lamblin Could you please install [PyCuda 2016.1.2](https://pypi.python.org/pypi/pycuda) in the same environment with Theano and Python 3?
If I remove PyCuda from my build I don't get the error. So if you don't get any error after installing PyCuda then I guess it's just something in the way I compile things. Thanks!

**Comment by nouiz** [comment]
Can you confirm the error you are having now is still the original error? Can you post a new version in case there is small difference in the stack trace that help us? Make sure to update Theano and libgpuarray to 0.9rc3 that we did, in case this is a different error related to what is already fixed.

**Comment by pnavarre** [comment]
I am sorry, I confirm that this was my wrongdoing. So as far as I'm concerned this issue can be closed.

Basically, I created two environments, one for Python 2 and another for Python 3. First, I compiled the source of Pycuda from source using Python2's environment. Later, I entered the same directory, I didn't clean the compiled source, and using Python3's environment I compiled and install. Obviously a bad practice which went unnoticed for long time. I fixed it by compiling for different environments in different directories.

Sorry and thank you for your support!
